File: src/main/java/io/wcygan/algorithms/sorting/QuickSort.java
package io.wcygan.algorithms.sorting;

/**
 * Implements QuickSort adapted from CLRS Chapter 7
 *
 * @param <T> the type of the values to sort
 */
public class QuickSort<T extends Comparable<T>> implements Sorter<T> {
  // TODO implement a quickcheck test that generates arrays of random sizes

  @Override
  public void sort(T[] arr) {
    quicksort(arr, 0, arr.length - 1);
  }

  private void quicksort(T[] arr, int p, int r) {
    if (p < r) {
      int q = partition(arr, p, r);
      quicksort(arr, p, q - 1);
      quicksort(arr, q + 1, r);
    }
  }

  private int partition(T[] arr, int p, int r) {
    T x = arr[r];
    int i = p - 1;
    for (int j = p; j < r; j++) {

      if (arr[j].compareTo(x) <= 0) {
        i = i + 1;
        swap(arr, i, j);
      }
    }
    swap(arr, i + 1, r);
    return i + 1;
  }

  private void swap(T[] arr, int first, int second) {
    T a = arr[first];
    arr[first] = arr[second];
    arr[second] = a;
  }
}

File: scala-practice/lib/src/io/wcygan/algorithms/sort/QuickSort.scala
package io.wcygan.algorithms.sort

object QuickSort extends Sorter {
  override def sort[T: Ordering](items: IndexedSeq[T]): IndexedSeq[T] = {
    if (items.length <= 1) items
    else {
      val pivot = items(items.length / 2)
      IndexedSeq.concat(
        sort(items.filter(Ordering[T].lt(_, pivot))),
        items.filter(Ordering[T].equiv(_, pivot)),
        sort(items.filter(Ordering[T].gt(_, pivot)))
      )
    }
  }
}
